apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -128,21 +137,34 @@ where
             // valid because of mempool check
             TryInto::<DateTimeUtc>::try_into(block_time).ok()
         });
+        let mut temp_block_gas_meter = BlockGasMeter::new(
+            self.wl_storage
+                .read(&parameters::storage::get_max_block_gas_key())
+                .expect("Error while reading from storage")
+                .expect("Missing max_block_gas parameter in storage"),
+        );
+        let mut temp_wl_storage = TempWlStorage::new(&self.wl_storage.storage);
+        let gas_table: BTreeMap<String, u64> = self
+            .wl_storage
+            .read(&parameters::storage::get_gas_table_storage_key())
+            .expect("Error while reading from storage")
+            .expect("Missing gas table in storage");
+        let mut vp_wasm_cache = self.vp_wasm_cache.clone();
+        let mut tx_wasm_cache = self.tx_wasm_cache.clone();
+
         let txs = txs
             .iter()
             .filter_map(|tx_bytes| {
-                if let Ok(tx) = Tx::try_from(tx_bytes.as_slice()) {
-                    // If tx doesn't have an expiration it is valid. If time cannot be
-                    // retrieved from block default to last block datetime which has
-                    // already been checked by mempool_validate, so it's valid
-                    if let (Some(block_time), Some(exp)) = (block_time.as_ref(), &tx.expiration) {
-                        if block_time > exp { return None }
-                    }
-                    if let Ok(TxType::Wrapper(_)) = process_tx(tx) {
-                        return Some(tx_bytes.clone());
+                match self.validate_wrapper_bytes(tx_bytes, &mut temp_block_gas_meter, block_time, &mut temp_wl_storage, &gas_table, &mut vp_wasm_cache, &mut tx_wasm_cache) {
```

Comment:
The gas check inside `validate_wrapper_bytes` could actually be extracted from there and moved into the `BlockSpaceAllocator`, effectively transforming it into a `BlockSpaceGasAllocator`. It would probably make for a more coherent logic but would not improve performance. Actually, given that the allocator stops at the first element of the iterator that cannot be placed in the block, we could end up with blocks containing less transactions